lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb

```diff
@@ -0,0 +1,202 @@
+          get_service_instances(connection, query_params,
+                                :on_premise        => true,
+                                :receptor_receiver => receiver,
+                                :receptor_params   => receptor_params)
+        end
+
+        public
+
+        def async_save_inventory(refresh_state_uuid, parser)
+          refresh_state_part_collected_at = Time.now.utc
+          refresh_state_part_uuid = SecureRandom.uuid
+          save_inventory(parser.collections.values, inventory_name, schema_name, refresh_state_uuid, refresh_state_part_uuid, refresh_state_part_collected_at)
+        end
+
+        def async_collecting_finished(entity_type, refresh_state_uuid, total_parts)
+          logger.info("ServiceInstance#refresh: finished collecting of #{entity_type} (#{refresh_state_uuid}). Total parts: #{total_parts}")
+        end
+
+        private
+
+        def refresh_part_cloud(tasks_id, refresh_state_uuid, refresh_state_part_uuid, query_params)
+          parser = TopologicalInventory::AnsibleTower::Parser.new(:tower_url => tower_hostname)
+
+          get_service_instances(connection, query_params).each do |service_instance|
+            parser.parse_service_instance(service_instance)
+          end
+
+          # Sending to Ingress API
+          logger.info("ServiceInstance#refresh - Task[ id: #{tasks_id} ] Sending to Ingress API...")
+          save_inventory(parser.collections.values, inventory_name, schema_name, refresh_state_uuid, refresh_state_part_uuid)
+          logger.info("ServiceInstance#refresh - Task[ id: #{tasks_id} ] Sending to Ingress API...Complete")
+        end
+
+        def connection
+          @connection ||= begin
+                            tower_user = authentication.username unless on_premise?
+                            tower_passwd = authentication.password unless on_premise?
+
+                            connection_manager.connect(
+                              :base_url       => tower_hostname,
+                              :username       => tower_user,
+                              :password       => tower_passwd,
+                              :receptor_node  => endpoint.receptor_node.to_s.strip,
+                              :account_number => account_number
+                            )
+                          end
+        end
+
+        def on_premise?
+          @on_premise ||= endpoint.receptor_node.to_s.strip.present?
+        end
+
+        def set_connection_data!
+          raise TopologicalInventory::Providers::Common::Operations::Source::ERROR_MESSAGES[:endpoint_not_found] unless endpoint
+          raise TopologicalInventory::Providers::Common::Operations::Source::ERROR_MESSAGES[:authentication_not_found] unless authentication
+
+          self.tower_hostname = full_hostname(endpoint)
+        end
+
+        # TODO: Join with other operations
```

Comment:
Yeah I think we could make use of the `Source` class in providers-common here, if we made the endpoint/auth methods public, would be nice to store the state there, definitely a follow-up though when we refactor a lot of the availability checking stuff like in #112.